File: Sources/Lib/Audio/LockFreeRingBuffer.swift
// Single-producer / single-consumer lock-free primitives used by the
// audio thread. Two shapes:
//
//   * `SPSCAudioRingBuffer` — power-of-two ring of `Float` samples with
//     two access patterns:
//       - **Consume style** (`write` + `consume`): the consumer
//         advances a read cursor; each sample is delivered exactly
//         once. Used by the CoreAudio capture and playback paths.
//       - **Snapshot style** (`appendConvertingDoubleToFloat` +
//         `readLatest`): the consumer takes the most-recent N
//         samples without advancing any cursor; the same window
//         can be re-read for FFTs at different lengths. Used by
//         `SpectrumAnalyzer`.
//     The two patterns share the same producer index, so a single
//     ring can serve either role — they don't mix on a given ring,
//     but the same primitive covers both.
//
//   * `SPSCQueue<T>` — generic SPSC FIFO queue. Used to pass
//     `AudioChunk` values between the capture, processing, and
//     playback threads inside `DSPEngineCore` without taking an
//     `NSLock`.
//
//   * `AtomicDouble` — a wait-free `Double` atom built on
//     `Atomic<UInt64>` round-tripped through the IEEE-754 bit pattern.
//     Used by the rate-adjust loop to publish the resampler ratio
//     from the playback thread to the processing thread.
//
// Real-time discipline
// --------------------
// All hot-path methods are wait-free, allocation-free, and free of
// Swift-runtime calls that could block. The producer always succeeds
// — if the consumer is so far behind that the buffer is full, the
// oldest unread data is silently overwritten (matching the original
// lock-based design's drop-on-overflow behaviour).

import Accelerate
import Synchronization

// MARK: - SPSC ring buffer

/// Lock-free SPSC ring buffer of `Float` samples. Power-of-two
/// capacity so wrap-around is a single bitmask. Producer publishes
/// new samples with a `release-store` on `writeIndex`; consumers
/// observe with an `acquire-load`, establishing happens-before
/// without locks.
///
/// Two consumer styles, both supported on the same instance — but
/// don't mix them on a single ring:
///
///   - **Consume:** call `consume(into:count:)` to drain samples.
///     Each sample is delivered to exactly one consumer call.
///     Used by `CoreAudioCapture` / `CoreAudioPlayback`.
///   - **Snapshot:** call `readLatest(into:count:)` to copy the
///     most-recent `count` samples without advancing any cursor.
///     The same samples can be re-read across calls. Used by
///     `SpectrumAnalyzer` to feed FFTs at different lengths.
public final class SPSCAudioRingBuffer: @unchecked Sendable {

  /// Capacity in samples (always a power of two).
  public let capacity: Int

  private let mask: Int
  private let storage: UnsafeMutableBufferPointer<Float>
  private let base: UnsafeMutablePointer<Float>
  /// Monotonically increasing count of samples written since
  /// allocation. The release-store synchronises with consumers'
  /// acquire-loads, so any reader that observes the new count is
  /// guaranteed to see the corresponding sample writes.
  private let writeIndex = Atomic<UInt64>(0)
  /// Monotonic samples drained by the consumer in `consume(...)`.
  /// Only used by the consume-style API; snapshot readers ignore
  /// this entirely.
  private let readIndex = Atomic<UInt64>(0)

  public init(minimumCapacity: Int) {
    let cap = SPSCAudioRingBuffer.roundUpToPowerOfTwo(Swift.max(2, minimumCapacity))
    self.capacity = cap
    self.mask = cap - 1
    let buffer = UnsafeMutableBufferPointer<Float>.allocate(capacity: cap)
    buffer.initialize(repeating: 0)
    self.storage = buffer
    guard let basePtr = buffer.baseAddress else {
      fatalError("Failed to allocate storage for SPSCAudioRingBuffer")
    }
    self.base = basePtr
  }

  deinit {
    storage.deallocate()
  }

  /// Total samples written since allocation. Observed with
  /// `.relaxed`; callers that need happens-before with the
  /// payload should use `consume` or `readLatest` instead.
  public var totalSamplesWritten: UInt64 {
    writeIndex.load(ordering: .relaxed)
  }

  /// Number of samples currently waiting to be consumed (for
  /// consume-style use). Always non-negative.
  public var availableToRead: Int {
    let w = writeIndex.load(ordering: .acquiring)
    let r = readIndex.load(ordering: .relaxed)
    return Int(w &- r)
  }

  // MARK: Producer

  /// **Producer-only.** Write `count` `Float` samples from
  /// `source` into the ring. `stride` lets the producer pull a
  /// single channel out of an interleaved buffer (`stride =
  /// channels`); pass `1` for non-interleaved input. Always
  /// succeeds — if the consumer is too far behind the oldest
  /// unread data is silently overwritten.
  public func write(source: UnsafePointer<Float>, count: Int, stride: Int = 1) {
    guard count > 0 else { return }
    var src = source
    var cnt = count
    if cnt > capacity {
      let skip = cnt - capacity
      src += skip * stride
      cnt = capacity
    }
    let w = writeIndex.load(ordering: .relaxed)
    let writeOffset = Int(w & UInt64(mask))
    let firstChunk = Swift.min(capacity - writeOffset, cnt)
    if stride == 1 {
      (base + writeOffset).update(from: src, count: firstChunk)
      if firstChunk < cnt {
        base.update(
          from: src + firstChunk,
          count: cnt - firstChunk)
      }
    } else {
      // Strided copy: extract every `stride`-th element of `source`
      // into the contiguous ring slot. `vDSP_vsadd` with a zero
      // scalar is a stride-aware copy — there's no dedicated
      // strided memcpy in vDSP.
      var zero: Float = 0
      vDSP_vsadd(
        src, vDSP_Stride(stride), &zero,
        base + writeOffset, 1,
        vDSP_Length(firstChunk))
      if firstChunk < cnt {
        vDSP_vsadd(
          src + (stride * firstChunk), vDSP_Stride(stride), &zero,
          base, 1,
          vDSP_Length(cnt - firstChunk))
      }
    }
    writeIndex.store(w &+ UInt64(cnt), ordering: .releasing)
  }

  /// **Producer-only.** Convert `count` `Double` samples from
  /// `source` to `Float` in a single `vDSP_vdpsp` call and write
  /// into the ring. Used by the spectrum-analyzer tap, which feeds
  /// engine-precision `Double` samples into a half-precision ring
  /// to halve memory.
  public func appendConvertingDoubleToFloat(_ source: UnsafePointer<Double>, count: Int) {
    guard count > 0 else { return }
    var src = source
    var cnt = count
    if cnt > capacity {
      let skip = cnt - capacity
      src += skip
      cnt = capacity
    }
    let w = writeIndex.load(ordering: .relaxed)
    let writeOffset = Int(w & UInt64(mask))
    let firstChunk = Swift.min(capacity - writeOffset, cnt)
    // vDSP_vdpsp: convert and store Double→Float, no allocation.
    vDSP_vdpsp(src, 1, base + writeOffset, 1, vDSP_Length(firstChunk))
    if firstChunk < cnt {
      let remaining = cnt - firstChunk
      vDSP_vdpsp(src + firstChunk, 1, base, 1, vDSP_Length(remaining))
    }
    writeIndex.store(w &+ UInt64(cnt), ordering: .releasing)
  }

  /// **Producer-only.** Write `count` zeros into the ring.
  /// Always succeeds — if the consumer is too far behind the oldest
  /// unread data is silently overwritten.
  public func writeSilence(count: Int) {
    guard count > 0 else { return }
    var cnt = count
    if cnt > capacity {
      cnt = capacity
    }
    let w = writeIndex.load(ordering: .relaxed)
    let writeOffset = Int(w & UInt64(mask))
    let firstChunk = Swift.min(capacity - writeOffset, cnt)

    (base + writeOffset).update(repeating: 0, count: firstChunk)
    if firstChunk < cnt {
      base.update(repeating: 0, count: cnt - firstChunk)
    }
    writeIndex.store(w &+ UInt64(cnt), ordering: .releasing)
  }

  // MARK: Consumer (consume style)

  /// **Consumer-only.** Copy up to `count` samples into `dest` and
  /// advance the read cursor. Returns the number of samples
  /// actually copied — may be less than `count` if fewer are
  /// available, in which case the remainder of `dest` is left
  /// untouched and the caller should fill it with silence.
  @discardableResult
  public func consume(into dest: UnsafeMutablePointer<Float>, count: Int) -> Int {
    guard count > 0 else { return 0 }
    var r = readIndex.load(ordering: .relaxed)
    let w = writeIndex.load(ordering: .acquiring)
    let avail = Int(w &- r)

    if avail > capacity {
      // Producer has overwritten unread data. Advance read pointer to the
      // oldest valid data (writeIndex - capacity).
      r = w - UInt64(capacity)
      readIndex.store(r, ordering: .releasing)
    }

    let n = Swift.min(Int(w &- r), count)
    guard n > 0 else { return 0 }
    let readOffset = Int(r & UInt64(mask))
    let firstChunk = Swift.min(capacity - readOffset, n)
    dest.update(from: base + readOffset, count: firstChunk)
    if firstChunk < n {
      (dest + firstChunk).update(
        from: base,
        count: n - firstChunk)
    }
    readIndex.store(r &+ UInt64(n), ordering: .releasing)
    return n
  }

  /// **Consumer-only.** Discard any pending samples without
  /// copying. Useful when the consumer wants to re-sync after a
  /// long stall.
  public func drain() {
    let w = writeIndex.load(ordering: .acquiring)
    readIndex.store(w, ordering: .releasing)
  }

  // MARK: Consumer (snapshot style)

  /// **Consumer.** Copy the most recent `count` samples into
  /// `dest` *without* advancing any cursor — subsequent calls
  /// can re-read overlapping windows. Returns `false` (without
  /// writing to `dest`) when fewer than `count` samples have been
  /// written so far.
  ///
  /// Tearing: in principle the producer can wrap the entire buffer
  /// during the consumer's memcpy. With the spectrum analyzer's
  /// 262 144-sample buffer at 48 kHz that's ~5.5 s of audio
  /// headroom — orders of magnitude longer than the consumer
  /// takes — so the snapshot is effectively atomic and we don't
  /// pay for a seqlock retry loop.
  public func readLatest(into dest: UnsafeMutablePointer<Float>, count: Int) -> Bool {
    guard count > 0, count <= capacity else { return false }
    let written = writeIndex.load(ordering: .acquiring)
    guard written >= UInt64(count) else { return false }
    let endIdx = Int(written & UInt64(mask))
    let startIdx = (endIdx + capacity - count) & mask
    let firstChunk = Swift.min(capacity - startIdx, count)
    dest.update(from: base + startIdx, count: firstChunk)
    if firstChunk < count {
      (dest + firstChunk).update(from: base, count: count - firstChunk)
    }
    return true
  }

  static func roundUpToPowerOfTwo(_ n: Int) -> Int {
    var v = 1
    while v < n { v <<= 1 }
    return v
  }
}

// MARK: - SPSC chunk queue

/// Lock-free single-producer / single-consumer FIFO queue of values of
/// arbitrary type `T`. Used to pass `AudioChunk` values between the
/// capture, processing, and playback threads inside `DSPEngineCore`
/// without taking an `NSLock`.
///
/// Power-of-two capacity. Slots store `Optional<T>` so the consumer
/// can clear back to `nil` on dequeue, dropping the ARC retain on
/// any class fields the value contains.
public final class SPSCQueue<T>: @unchecked Sendable {

  public let capacity: Int

  private let mask: Int
  /// Slot storage. Each slot holds `nil` when empty; the producer
  /// fills it on enqueue and the consumer clears it back to `nil` on
  /// dequeue (so any `T` reference fields drop their ARC retain when
  /// the consumer is done).
  private let storage: UnsafeMutableBufferPointer<T?>
  private let writeIndex = Atomic<UInt64>(0)
  private let readIndex = Atomic<UInt64>(0)

  public init(minimumCapacity: Int) {
    let cap = SPSCAudioRingBuffer.roundUpToPowerOfTwo(Swift.max(2, minimumCapacity))
    self.capacity = cap
    self.mask = cap - 1
    let buf = UnsafeMutableBufferPointer<T?>.allocate(capacity: cap)
    buf.initialize(repeating: nil)
    self.storage = buf
  }

  deinit {
    // Releasing each slot drops the ARC retain on any contained
    // references; deinitialize then deallocate the raw storage.
    for i in 0..<storage.count { storage[i] = nil }
    storage.deinitialize()
    storage.deallocate()
  }

  /// Number of currently-queued items. Approximate when read from a
  /// thread that is neither the producer nor the consumer.
  public var count: Int {
    let w = writeIndex.load(ordering: .acquiring)
    let r = readIndex.load(ordering: .relaxed)
    return Int(w &- r)
  }

  /// **Producer-only.** Append `value`; returns `false` (without
  /// storing it) when the queue is at capacity. The caller decides
  /// what to do — drop, log, or retry.
  public func enqueue(_ value: sending T) -> Bool {
    let w = writeIndex.load(ordering: .relaxed)
    let r = readIndex.load(ordering: .acquiring)
    if w &- r >= UInt64(capacity) { return false }
    storage[Int(w & UInt64(mask))] = value
    writeIndex.store(w &+ 1, ordering: .releasing)
    return true
  }

  /// **Consumer-only.** Pop the next item; returns `nil` when empty.
  public func dequeue() -> sending T? {
    let r = readIndex.load(ordering: .relaxed)
    let w = writeIndex.load(ordering: .acquiring)
    if r == w { return nil }
    let slot = Int(r & UInt64(mask))
    let value = storage[slot]
    storage[slot] = nil
    readIndex.store(r &+ 1, ordering: .releasing)
    return value
  }

  /// **Consumer-only.** Discard all queued items.
  public func drain() {
    while dequeue() != nil {}
  }
}

// MARK: - Atomic Double via UInt64 bit-pattern

/// Lock-free atomic `Double`. Swift's `Atomic` doesn't directly support
/// floating-point payloads, so we round-trip through the IEEE-754 bit
/// pattern via `UInt64`. Aligned 64-bit loads and stores are atomic on
/// every platform Swift 6 runs on, so this is genuinely wait-free.
///
/// Implicitly `Sendable` — the only stored property is `Atomic<UInt64>`,
/// which is itself `Sendable`, so the compiler infers the conformance
/// without `@unchecked`.
public final class AtomicDouble: Sendable {
  private let bits: Atomic<UInt64>

  public init(_ value: Double) {
    self.bits = Atomic<UInt64>(value.bitPattern)
  }

  public var value: Double {
    get { Double(bitPattern: bits.load(ordering: .acquiring)) }
    set { bits.store(newValue.bitPattern, ordering: .releasing) }
  }
}

File: Sources/CDSP/Audio/lock_free_ring_buffer.c
// Single-producer / single-consumer lock-free primitives used by the audio
// thread.
#include "Audio/lock_free_ring_buffer.h"

#include <stdlib.h>

struct spsc_audio_ring_buffer {
  size_t capacity;
  size_t mask;
  float* storage;
  _Atomic uint64_t write_index;
  _Atomic uint64_t read_index;
};

struct spsc_queue {
  size_t capacity;
  size_t mask;
  void** storage;
  _Atomic uint64_t write_index;
  _Atomic uint64_t read_index;
};

uint64_t spsc_audio_ring_buffer_get_total_samples_written(
    const spsc_audio_ring_buffer_t* ring) {
  return atomic_load_explicit(&ring->write_index, memory_order_relaxed);
}

size_t spsc_audio_ring_buffer_get_available_to_read(
    const spsc_audio_ring_buffer_t* ring) {
  // Acquire barrier on write_index ensures that any data written by the
  // producer prior to updating write_index is visible to this thread.
  uint64_t w = atomic_load_explicit(&ring->write_index, memory_order_acquire);
  // Relaxed load is sufficient for read_index as it is only modified by the
  // consumer (which is typically the calling thread of this function).
  uint64_t r = atomic_load_explicit(&ring->read_index, memory_order_relaxed);
  // Unsigned subtraction correctly handles overflow wrap-around.
  return (size_t)(w - r);
}

size_t spsc_audio_ring_buffer_get_available_to_write(
    const spsc_audio_ring_buffer_t* ring) {
  // Relaxed load is sufficient for write_index as it is only modified by the
  // producer (which is typically the calling thread).
  uint64_t w = atomic_load_explicit(&ring->write_index, memory_order_relaxed);
  // Acquire barrier on read_index ensures we see the consumer's latest read
  // index before we decide how much space we can write to, avoiding overwriting
  // unread data.
  uint64_t r = atomic_load_explicit(&ring->read_index, memory_order_acquire);
  size_t occupied = (size_t)(w - r);
  if (occupied >= ring->capacity) {
    return 0;
  }
  return ring->capacity - occupied;
}

size_t spsc_audio_ring_buffer_get_capacity(
    const spsc_audio_ring_buffer_t* ring) {
  return ring ? ring->capacity : 0;
}

size_t spsc_queue_get_count(const spsc_queue_t* queue) {
  uint64_t w = atomic_load_explicit(&queue->write_index, memory_order_acquire);
  uint64_t r = atomic_load_explicit(&queue->read_index, memory_order_relaxed);
  return (size_t)(w - r);
}

size_t spsc_queue_get_capacity(const spsc_queue_t* queue) {
  return queue ? queue->capacity : 0;
}
#include <string.h>

#ifdef ENABLE_ACCELERATE
#include <Accelerate/Accelerate.h>
#endif

// MARK: - SPSCAudioRingBuffer Implementation

spsc_audio_ring_buffer_t* spsc_audio_ring_buffer_create(
    size_t minimum_capacity) {
  size_t cap = spsc_audio_ring_buffer_round_up_to_power_of_two(
      minimum_capacity < 2 ? 2 : minimum_capacity);
  spsc_audio_ring_buffer_t* ring =
      (spsc_audio_ring_buffer_t*)malloc(sizeof(spsc_audio_ring_buffer_t));
  if (!ring) return NULL;
  ring->capacity = cap;
  ring->mask = cap - 1;
  ring->storage = (float*)calloc(cap, sizeof(float));
  if (!ring->storage) {
    free(ring);
    return NULL;
  }
  atomic_init(&ring->write_index, 0);
  atomic_init(&ring->read_index, 0);
  return ring;
}

void spsc_audio_ring_buffer_free(spsc_audio_ring_buffer_t* ring) {
  if (!ring) return;
  free(ring->storage);
  free(ring);
}

void spsc_audio_ring_buffer_write(spsc_audio_ring_buffer_t* ring,
                                  const float* source, size_t count,
                                  size_t stride) {
  if (count == 0 || !ring || !source) return;
  const float* src = source;
  size_t cnt = count;
  // If count exceeds capacity, only write the most recent 'capacity' elements.
  if (cnt > ring->capacity) {
    size_t skip = cnt - ring->capacity;
    src += skip * stride;
    cnt = ring->capacity;
  }
  // Relaxed load: we are the only writer.
  uint64_t w = atomic_load_explicit(&ring->write_index, memory_order_relaxed);
  size_t write_offset = (size_t)(w & ring->mask);

  // Calculate how many samples we can write before wrapping around the end of
  // the ring.
  size_t first_chunk = ring->capacity - write_offset;
  if (first_chunk > cnt) first_chunk = cnt;

  if (stride == 1) {
    memcpy(ring->storage + write_offset, src, first_chunk * sizeof(float));
    if (first_chunk < cnt) {
      memcpy(ring->storage, src + first_chunk,
             (cnt - first_chunk) * sizeof(float));
    }
  } else {
#ifdef ENABLE_ACCELERATE
    // Strided copy: extract every `stride`-th element of `source`
    // into the contiguous ring slot. `vDSP_vsadd` with a zero
    // scalar is used as a stride-aware copy because vDSP lacks a
    // dedicated strided memcpy.
    float zero = 0.0f;
    vDSP_vsadd(src, stride, &zero, ring->storage + write_offset, 1,
               first_chunk);
    if (first_chunk < cnt) {
      vDSP_vsadd(src + (stride * first_chunk), stride, &zero, ring->storage, 1,
                 cnt - first_chunk);
    }
#else
    for (size_t i = 0; i < first_chunk; i++) {
      ring->storage[write_offset + i] = src[i * stride];
    }
    if (first_chunk < cnt) {
      for (size_t i = 0; i < cnt - first_chunk; i++) {
        ring->storage[i] = src[(first_chunk + i) * stride];
      }
    }
#endif
  }
  // Release store: makes all previous memory writes (the data copy) visible to
  // the reader before they see the updated write_index.
  atomic_store_explicit(&ring->write_index, w + cnt, memory_order_release);
}

void spsc_audio_ring_buffer_append_converting_double_to_float(
    spsc_audio_ring_buffer_t* ring, const double* source, size_t count) {
  if (count == 0 || !ring || !source) return;
  const double* src = source;
  size_t cnt = count;
  if (cnt > ring->capacity) {
    size_t skip = cnt - ring->capacity;
    src += skip;
    cnt = ring->capacity;
  }
  uint64_t w = atomic_load_explicit(&ring->write_index, memory_order_relaxed);
  size_t write_offset = (size_t)(w & ring->mask);
  size_t first_chunk = ring->capacity - write_offset;
  if (first_chunk > cnt) first_chunk = cnt;

#ifdef ENABLE_ACCELERATE
  // vDSP_vdpsp: convert and store Double->Float, no allocation.
  vDSP_vdpsp(src, 1, ring->storage + write_offset, 1, first_chunk);
  if (first_chunk < cnt) {
    vDSP_vdpsp(src + first_chunk, 1, ring->storage, 1, cnt - first_chunk);
  }
#else
  for (size_t i = 0; i < first_chunk; i++) {
    ring->storage[write_offset + i] = (float)src[i];
  }
  if (first_chunk < cnt) {
    for (size_t i = 0; i < cnt - first_chunk; i++) {
      ring->storage[i] = (float)src[first_chunk + i];
    }
  }
#endif
  atomic_store_explicit(&ring->write_index, w + cnt, memory_order_release);
}

void spsc_audio_ring_buffer_write_silence(spsc_audio_ring_buffer_t* ring,
                                          size_t count) {
  if (count == 0 || !ring) return;
  size_t cnt = count;
  if (cnt > ring->capacity) {
    cnt = ring->capacity;
  }
  uint64_t w = atomic_load_explicit(&ring->write_index, memory_order_relaxed);
  size_t write_offset = (size_t)(w & ring->mask);
  size_t first_chunk = ring->capacity - write_offset;
  if (first_chunk > cnt) first_chunk = cnt;

  memset(ring->storage + write_offset, 0, first_chunk * sizeof(float));
  if (first_chunk < cnt) {
    memset(ring->storage, 0, (cnt - first_chunk) * sizeof(float));
  }
  atomic_store_explicit(&ring->write_index, w + cnt, memory_order_release);
}

size_t spsc_audio_ring_buffer_consume(spsc_audio_ring_buffer_t* ring,
                                      float* dest, size_t count) {
  if (count == 0 || !ring || !dest) return 0;
  // Relaxed load: we are the only reader.
  uint64_t r = atomic_load_explicit(&ring->read_index, memory_order_relaxed);
  // Acquire barrier: ensure we see the writes to the buffer elements before we
  // read them.
  uint64_t w = atomic_load_explicit(&ring->write_index, memory_order_acquire);
  size_t avail = (size_t)(w - r);

  if (avail > ring->capacity) {
    // Producer has overwritten unread data. This can happen if the producer
    // writes faster than we consume and the buffer wraps around. Advance read
    // pointer to the oldest valid data (write_index - capacity).
    r = w - (uint64_t)ring->capacity;
    atomic_store_explicit(&ring->read_index, r, memory_order_release);
    avail = ring->capacity;
  }

  size_t n = avail < count ? avail : count;
  if (n == 0) return 0;

  // Calculate read offset and handle wrap-around.
  size_t read_offset = (size_t)(r & ring->mask);
  size_t first_chunk = ring->capacity - read_offset;
  if (first_chunk > n) first_chunk = n;

  memcpy(dest, ring->storage + read_offset, first_chunk * sizeof(float));
  if (first_chunk < n) {
    memcpy(dest + first_chunk, ring->storage,
           (n - first_chunk) * sizeof(float));
  }
  // Release store: let the producer know we've freed up slot space up to r + n.
  atomic_store_explicit(&ring->read_index, r + n, memory_order_release);
  return n;
}

size_t spsc_audio_ring_buffer_consume_stride(spsc_audio_ring_buffer_t* ring,
                                             float* dest, size_t count,
                                             size_t stride) {
  if (count == 0 || !ring || !dest) return 0;
  // Relaxed load: we are the only reader.
  uint64_t r = atomic_load_explicit(&ring->read_index, memory_order_relaxed);
  // Acquire barrier: ensure we see the writes to the buffer elements before we
  // read them.
  uint64_t w = atomic_load_explicit(&ring->write_index, memory_order_acquire);
  size_t avail = (size_t)(w - r);

  if (avail > ring->capacity) {
    // Producer has overwritten unread data. This can happen if the producer
    // writes faster than we consume and the buffer wraps around. Advance read
    // pointer to the oldest valid data (write_index - capacity).
    r = w - (uint64_t)ring->capacity;
    atomic_store_explicit(&ring->read_index, r, memory_order_release);
    avail = ring->capacity;
  }

  size_t n = avail < count ? avail : count;
  if (n == 0) return 0;

  // Calculate read offset and handle wrap-around.
  size_t read_offset = (size_t)(r & ring->mask);
  size_t first_chunk = ring->capacity - read_offset;
  if (first_chunk > n) first_chunk = n;

  if (stride == 1) {
    memcpy(dest, ring->storage + read_offset, first_chunk * sizeof(float));
    if (first_chunk < n) {
      memcpy(dest + first_chunk, ring->storage,
             (n - first_chunk) * sizeof(float));
    }
  } else {
#ifdef ENABLE_ACCELERATE
    float zero = 0.0f;
    vDSP_vsadd(ring->storage + read_offset, 1, &zero, dest, stride,
               first_chunk);
    if (first_chunk < n) {
      vDSP_vsadd(ring->storage, 1, &zero, dest + (stride * first_chunk), stride,
                 n - first_chunk);
    }
#else
    for (size_t i = 0; i < first_chunk; i++) {
      dest[i * stride] = ring->storage[read_offset + i];
    }
    if (first_chunk < n) {
      for (size_t i = 0; i < n - first_chunk; i++) {
        dest[(first_chunk + i) * stride] = ring->storage[i];
      }
    }
#endif
  }
  // Release store: let the producer know we've freed up slot space up to r + n.
  atomic_store_explicit(&ring->read_index, r + n, memory_order_release);
  return n;
}

void spsc_audio_ring_buffer_drain(spsc_audio_ring_buffer_t* ring) {
  if (!ring) return;
  uint64_t w = atomic_load_explicit(&ring->write_index, memory_order_acquire);
  atomic_store_explicit(&ring->read_index, w, memory_order_release);
}

bool spsc_audio_ring_buffer_read_latest_at(const spsc_audio_ring_buffer_t* ring,
                                           float* dest, size_t count,
                                           uint64_t written) {
  if (!ring || !dest || count == 0 || count > ring->capacity) return false;
  if (written < (uint64_t)count) return false;

  uint64_t written_before =
      atomic_load_explicit(&ring->write_index, memory_order_acquire);
  if (written_before - written > (uint64_t)(ring->capacity - count)) {
    return false;
  }

  size_t end_idx = (size_t)(written & ring->mask);
  size_t start_idx = (end_idx + ring->capacity - count) & ring->mask;
  size_t first_chunk = ring->capacity - start_idx;
  if (first_chunk > count) first_chunk = count;

  memcpy(dest, ring->storage + start_idx, first_chunk * sizeof(float));
  if (first_chunk < count) {
    memcpy(dest + first_chunk, ring->storage,
           (count - first_chunk) * sizeof(float));
  }

  uint64_t written_after =
      atomic_load_explicit(&ring->write_index, memory_order_acquire);
  if (written_after - written > (uint64_t)(ring->capacity - count)) {
    return false;
  }
  return true;
}

bool spsc_audio_ring_buffer_read_latest(const spsc_audio_ring_buffer_t* ring,
                                        float* dest, size_t count) {
  if (!ring || !dest || count == 0 || count > ring->capacity) return false;

  int retries = 5;
  while (retries > 0) {
    uint64_t written =
        atomic_load_explicit(&ring->write_index, memory_order_acquire);
    if (written >= (uint64_t)count) {
      if (spsc_audio_ring_buffer_read_latest_at(ring, dest, count, written)) {
        return true;
      }
    } else {
      return false;
    }
    retries--;
  }
  return false;
}

// MARK: - SPSCQueue Implementation

spsc_queue_t* spsc_queue_create(size_t minimum_capacity) {
  size_t cap = spsc_audio_ring_buffer_round_up_to_power_of_two(
      minimum_capacity < 2 ? 2 : minimum_capacity);
  spsc_queue_t* queue = (spsc_queue_t*)malloc(sizeof(spsc_queue_t));
  if (!queue) return NULL;
  queue->capacity = cap;
  queue->mask = cap - 1;
  queue->storage = (void**)calloc(cap, sizeof(void*));
  if (!queue->storage) {
    free(queue);
    return NULL;
  }
  atomic_init(&queue->write_index, 0);
  atomic_init(&queue->read_index, 0);
  return queue;
}

void spsc_queue_free(spsc_queue_t* queue) {
  if (!queue) return;
  // Clearing each slot to NULL; deinitialize then deallocate the raw storage.
  free(queue->storage);
  free(queue);
}

bool spsc_queue_enqueue(spsc_queue_t* queue, void* value) {
  if (!queue) return false;
  uint64_t w = atomic_load_explicit(&queue->write_index, memory_order_relaxed);
  uint64_t r = atomic_load_explicit(&queue->read_index, memory_order_acquire);
  if (w - r >= (uint64_t)queue->capacity) return false;
  queue->storage[(size_t)(w & queue->mask)] = value;
  atomic_store_explicit(&queue->write_index, w + 1, memory_order_release);
  return true;
}

void* spsc_queue_dequeue(spsc_queue_t* queue) {
  if (!queue) return NULL;
  uint64_t r = atomic_load_explicit(&queue->read_index, memory_order_relaxed);
  uint64_t w = atomic_load_explicit(&queue->write_index, memory_order_acquire);
  if (r == w) return NULL;
  size_t slot = (size_t)(r & queue->mask);
  void* value = queue->storage[slot];
  queue->storage[slot] = NULL;
  atomic_store_explicit(&queue->read_index, r + 1, memory_order_release);
  return value;
}

void spsc_queue_drain(spsc_queue_t* queue) {
  if (!queue) return;
  while (spsc_queue_dequeue(queue) != NULL) {
  }
}

// MARK: - AtomicDouble Implementation

atomic_double_t* atomic_double_create(double value) {
  atomic_double_t* a = (atomic_double_t*)malloc(sizeof(atomic_double_t));
  if (!a) return NULL;
  atomic_double_init(a, value);
  return a;
}

void atomic_double_free(atomic_double_t* a) {
  if (!a) return;
  free(a);
}
